### PR TITLE
Centralize version data

### PR DESCRIPTION
**Pull request by levitte**
This moves all the version and release information to a file `VERSION` and parses that instead of `include/openssl/opensslv.h`, and also makes `include/openssl/opensslv.h` a generated file, which makes it easier to read as well as to parse.

### CODE REVIEW

**Comment by levitte** [comment]
I had *not* expected the shlibload test to fail because of this change...

**Comment by levitte** [comment]
CIs are happy

**Comment by mattcaswell** [comment]
Your temporary commit still appears in this PR

File: doc/internal/man7/VERSION.pod [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+=head2 Affected configuration data
+
+The following items in %config from F<configdata.pm> are affected:
+
+=over 4
+
+=item $config{major}, $config{minor}, $config{patch}, $config{shlib_version}
+
+These items get theird values from B<MAJOR>, B<MINOR>, B<PATCH>, and
```
Comment (mattcaswell) [inline doc/internal/man7/VERSION.pod]: Typo: their

File: doc/internal/man7/VERSION.pod [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+The minimal version number, a string composed from B<MAJOR>, B<MINOR> and
+B<PATCH>, separated by periods.  For C<MAJOR=3>, C<MINOR=0> and C<PATCH=0>,
+the string will be C<3.0.0>.
+
+=item $config{full_version}
+
+The fully loaded version number, a string composed from $config{version},
+$config{prerelease} and $config{build_metadata}.  For C<MAJOR=3>, C<MINOR=0>,
+C<PATCH=0>, B<PRE_RELEASE_TAG=dev>, B<BUILD_METADATA=>, the string will be
```
Comment (mattcaswell) [inline doc/internal/man7/VERSION.pod]: It might be helpful to show this example with non-empty metadata

File: include/openssl/opensslv.h.in [hunk 3]
```diff
@@ -86,43 +102,23 @@ extern "C" {
 # define OPENSSL_VERSION_PREREQ(maj,min)                                \
```
Comment (mattcaswell) [inline include/openssl/opensslv.h.in]: This is one of "newly added" undocumented symbols (see #9095). While you're doing version related stuff in this PR anyway - please could you document this?

File: include/openssl/opensslv.h.in [hunk 4]
```diff
@@ -86,43 +102,23 @@ extern "C" {
...
     ((OPENSSL_VERSION_MAJOR << 16) + OPENSSL_VERSION_MINOR >= ((maj) << 16) + (min))
 
-/* Helper macros for CPP string composition */
-#   define OPENSSL_MSTR_HELPER(x) #x
-#   define OPENSSL_MSTR(x) OPENSSL_MSTR_HELPER(x)
```
Comment (mattcaswell) [inline include/openssl/opensslv.h.in]: These are 2 of the other newly added undocumented symbols. Yay!

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: VERSION [hunk 5]
```diff
@@ -0,0 +1,7 @@
+MAJOR=3
+MINOR=0
+PATCH=0
```
Comment (levitte) [inline VERSION]: A question here is if these should be one line instead:

```
VERSION=3.0.0
```

Comment (levitte) [reply VERSION]: (Configure will split it it, of course)

File: doc/internal/man7/VERSION.pod (hunk 1 above)
Comment (levitte) [reply doc/internal/man7/VERSION.pod]: Fixed

File: doc/internal/man7/VERSION.pod (hunk 2 above)
Comment (levitte) [reply doc/internal/man7/VERSION.pod]: So, two examples then.

Comment (levitte) [reply doc/internal/man7/VERSION.pod]: Done

File: include/openssl/opensslv.h.in (hunk 3 above)
Comment (mattcaswell) [reply include/openssl/opensslv.h.in]: No, I don't mind that.

File: doc/internal/man7/VERSION.pod [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+=item $config{full_version}
+
+The fully loaded version number, a string composed from $config{version},
+$config{prerelease} and $config{build_metadata}.  See   See L</EXAMPLE> for
+a few examples.
+
+=back
+
+=head1 EXAMPLE
```
Comment (mattcaswell) [inline doc/internal/man7/VERSION.pod]: Should be `EXAMPLES`. I suspect doc-nits might complain about that.

**Review by mattcaswell — APPROVED** [review]
Approved assuming the nit is fixed and CIs agree

File: doc/internal/man7/VERSION.pod (hunk 6 above)
Comment (levitte) [reply doc/internal/man7/VERSION.pod]: Good point

**Comment by levitte** [comment]
Merged.

20551b2e62ed8c2fa97e0106a0350f40f3422dd5 Configure: get version from the file 'VERSION' instead of 'opensslv.h'
f3866324f0e9bd33feb48aefbe18d8f465d8f2a2 Generate include/openssl/opensslv.h
9495a6c45bc7cd69845f9a22eec1fbf8ac735b81 Remove the version number in README
f5453462117e3deb4b077756e2d24edb2ff3b1af Move the version function declarations to include/openssl/crypto.h

File: doc/internal/man7/VERSION.pod (hunk 6 above)
Comment (richsalz) [reply doc/internal/man7/VERSION.pod]: ```./util/find-doc-nits -n``` can usually avoid an extra CA-failure round trip.

**Comment by slontis** [comment]
@levitte: This change removed
/* Helper macros for CPP string composition */
#   define OPENSSL_MSTR_HELPER(x) #x
#   define OPENSSL_MSTR(x) OPENSSL_MSTR_HELPER(x)

If this was public before does it need to go back in?
There are a few references to this macro in the codebase (there could also be external ones)..

It looks like we need a travis entry for -enable-trace.

**Comment by t8m** [comment]
It was not public in a released version.

**Comment by levitte** [comment]
If they are needed, they can be re-added.  `include/openssl/macros.h` might be a good place for them.

**Comment by levitte** [comment]
PR welcome!

**Pull request closed without merging**